// @prismicio/types-internal/lib/import/converters/fields/nestable/RichText/textBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.textBlockConverter = void 0;
const Link_1 = require("../Link");
function textBlockConverter(block, assets) {
    return {
        type: block.type,
        content: {
            text: block.text,
            ...(() => {
                if (block.spans) {
                    const convertedSpans = spansConverter(block.spans, assets);
                    if (!convertedSpans)
                        return {};
                    return { spans: convertedSpans };
                }
                return {};
            })(),
        },
        ...(block.direction ? { direction: block.direction } : {}),
    };
}
exports.textBlockConverter = textBlockConverter;
function spansConverter(spans, assets) {
    return spans.reduce((acc, span) => {
        switch (span.type) {
            case "em":
            case "strong":
                return [...acc, span];
            case "label":
                return [
                    ...acc,
                    {
                        ...span,
                        data: span.data.label,
                    },
                ];
            case "hyperlink": {
                const convertedLink = (0, Link_1.linkConverter)(span.data, assets);
                if (!convertedLink)
                    return acc;
                return [
                    ...acc,
                    {
                        type: span.type,
                        start: span.start,
                        end: span.end,
                        data: convertedLink.value,
                    },
                ];
            }
        }
    }, []);
}
