// @slicemachine/manager/dist/index.js
import { createSliceMachineManager } from "./managers/createSliceMachineManager.js";
import { createSliceMachineManagerMiddleware } from "./managers/createSliceMachineManagerMiddleware.js";
import { createPrismicAuthManager } from "./auth/createPrismicAuthManager.js";
import { createPrismicAuthManagerMiddleware } from "./auth/createPrismicAuthManagerMiddleware.js";
import { InternalError, NotFoundError, PluginError, PluginHookResultError, SliceMachineError, UnauthenticatedError, UnauthorizedError, UnexpectedDataError } from "./errors.js";
import { DecodeError } from "./lib/DecodeError.js";
import { REPOSITORY_NAME_VALIDATION } from "./constants/REPOSITORY_NAME_VALIDATION.js";
export {
  DecodeError,
  InternalError,
  NotFoundError,
  PluginError,
  PluginHookResultError,
  REPOSITORY_NAME_VALIDATION,
  SliceMachineError,
  UnauthenticatedError,
  UnauthorizedError,
  UnexpectedDataError,
  createPrismicAuthManager,
  createPrismicAuthManagerMiddleware,
  createSliceMachineManager,
  createSliceMachineManagerMiddleware
};
