// @prismicio/types-internal/lib/import/validators/fields/nestable/Link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportLink = exports.Link = void 0;
const tslib_1 = require("tslib");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const DocumentId_1 = require("../../../../utils/DocumentId");
const validators_1 = require("../../../../validators");
const function_1 = require("../../../../validators/function");
const ImportContent_1 = require("../ImportContent");
const LinkTypeValidator = t.type({
    link_type: (0, function_1.withCustomError)(t.union([t.literal("Web"), t.literal("Document"), t.literal("Media")]), () => "The value must be `Web`, `Document` or `Media`"),
});
const WebLink = t.exact(t.intersection([
    t.type({
        link_type: t.literal("Web"),
        url: validators_1.String,
    }),
    t.partial({
        target: validators_1.String,
    }),
]));
const DocumentLink = t.strict({
    link_type: t.literal("Document"),
    id: DocumentId_1.DocumentId,
});
const MediaLink = t.strict({
    link_type: t.literal("Media"),
    id: validators_1.String,
});
exports.Link = LinkTypeValidator.pipe(t.union([WebLink, DocumentLink, MediaLink]));
// This is the default value for the link
const AnyLink = t.strict({
    link_type: t.literal("Any"),
});
exports.ImportLink = (0, ImportContent_1.ImportContent)("Link", (0, validators_1.DefaultOrElse)(AnyLink)(exports.Link));
